framework/src/service/ServiceRegistrationBase.cpp

```diff
@@ -209,28 +209,34 @@ namespace cppmicroservices
 
         CoreBundleContext* coreContext = nullptr;
 
+        bool isUnregistering(false); // expected state
+
+        if (!atomic_compare_exchange_strong(&d->coreInfo->unregistering, &isUnregistering, true))
+        {
+            // someone else is unregistering
+            return;
+        }
+
+        // d->coreInfo->unregistering is now true
+
+        // if unavailable
         if (!d->coreInfo->available)
         {
+            // reset isUnregistering to true
+            isUnregistering = true;
```

Comment:
I don't understand the logic of setting `isUnregistering` to `true`. If `d->coreInfo->available` is `false`, doesn't that mean that there is no unregistering taking place?

Reply:
if `d->coreInfo->available` is `false` then that means that unregistration has already happened. This means we should not unregister again at this time. Therefore, we need to exit this function using a throw. That being said, before that I am setting `d->coreInfo->unregistering` to false because we are no longer in this function unregistering. To do that, I first set `isUnregistering` to true as that is the expected value in `atomic_compare_exchange_strong`.

Reply:
if the unregistration already happened, then `isUnregistering` should be `false`, right? since the unregistration already happened and this code running on the thread is not performing any unregistration.

is an atomic compare and exchange necessary? or is setting the atomic `d->coreInfo->unregistering` variable enough?